toolchain/check/eval.cpp

```diff
@@ -585,39 +586,48 @@ static auto GetConstantFacetTypeInfo(EvalContext& eval_context,
                                      const SemIR::FacetTypeInfo& orig,
                                      Phase* phase) -> SemIR::FacetTypeInfo {
   SemIR::FacetTypeInfo info;
+
   info.extend_constraints.reserve(orig.extend_constraints.size());
   for (const auto& interface : orig.extend_constraints) {
     info.extend_constraints.push_back(
         {.interface_id = interface.interface_id,
          .specific_id =
              GetConstantValue(eval_context, interface.specific_id, phase)});
   }
+
   info.self_impls_constraints.reserve(orig.self_impls_constraints.size());
   for (const auto& interface : orig.self_impls_constraints) {
     info.self_impls_constraints.push_back(
         {.interface_id = interface.interface_id,
          .specific_id =
              GetConstantValue(eval_context, interface.specific_id, phase)});
   }
-  info.rewrite_constraints.reserve(orig.rewrite_constraints.size());
-  for (const auto& rewrite : orig.rewrite_constraints) {
+
+  // Rewrite constraints are resolved first before evaluation, so that we can
+  // work with the `ImplWitnessAccess` references to `.Self` without them
+  // evaluating to a value.
+  info.rewrite_constraints = orig.rewrite_constraints;
+  if (!ResolveFacetTypeRewriteConstraints(eval_context.context(), loc_id,
+                                          info.rewrite_constraints)) {
+    // TODO: Should ResolveFacetTypeRewriteConstraints() move into
+    // eval.cpp so it can set the Phase directly?
+
+    // FIXME: Is this needed anymore? We will see the errors below now, right?
+    *phase = Phase::UnknownDueToError;
+  }
+  for (auto& rewrite : info.rewrite_constraints) {
     // `where` requirements using `.Self` should not be considered symbolic.
     auto lhs_id =
         GetConstantValueIgnoringPeriodSelf(eval_context, rewrite.lhs_id, phase);
     auto rhs_id =
         GetConstantValueIgnoringPeriodSelf(eval_context, rewrite.rhs_id, phase);
-    info.rewrite_constraints.push_back({.lhs_id = lhs_id, .rhs_id = rhs_id});
+    rewrite = {.lhs_id = lhs_id, .rhs_id = rhs_id};
```

Comment:
I'm trying to puzzle through the assignment, why not:

```suggestion
    rewrite.lhs_id =
        GetConstantValueIgnoringPeriodSelf(eval_context, rewrite.lhs_id, phase);
    rewrite.rhs_id =
        GetConstantValueIgnoringPeriodSelf(eval_context, rewrite.rhs_id, phase);
```

Is this a holdover from the `rewrite_constraints.push_back`, or is there some issue with constant evaluation ordering (which might be worth a comment)?

Reply:
I did this then undid it because the assignment means if we add another field to the RewriteConstraint type (it's not an instruction) then the assignment version fails to build and points out the need to handle it here, whereas the per-field assignments don't.

Reply:
> if we add another field to the RewriteConstraint type (it's not an instruction) then the assignment version fails to build

Is that a bug, or a feature? What field are you planning on adding to `RewriteConstraint` that it would be okay to not rewrite (or rewrite)?

Reply:
Note, I wasn't intending this to block review. I was more in a comment-approve-merge frame of mind. It's every other unaddressed comment that's the issue.

Reply:
Much like an enum in a switch telling you when you need to handle a new case, I think it's a feature to have to think about new fields. This is similar to some other bugs we've seen/addressed by constructing new objects instead of constructing up front and mutating fields - I am thinking of something in the pattern matching that I think you reviewed.